plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs

```diff
@@ -116,33 +122,52 @@ function and the 'TypeScheme' of the built-in function will be derived automatic
 monomorphic and simply-polymorphic cases no types need to be specified at all.
 -}
 
+type family Length xs where
+    Length '[]       = 'Z
+    Length (_ ': xs) = 'S (Length xs)
+
 type family GetArgs a :: [GHC.Type] where
     GetArgs (a -> b) = a ': GetArgs b
     GetArgs _        = '[]
 
 -- | A class that allows us to derive a monotype for a builtin.
 class KnownMonotype val args res a | args res -> a, a -> res where
     knownMonotype :: TypeScheme val args res
+    knownMonoruntime :: RuntimeScheme (Length args)
+    toImmediateF :: FoldArgs args res -> ToDenotationType val (Length args)
+    toDeferredF :: ReadKnownM (FoldArgs args res) -> ToDenotationType val (Length args)
```

Comment:
I think this stuff probably needs a note (we'll hopefully delete it later but if we don't we'll wish we had the note).

Reply:
In particular, I think we need a short note just explaining that we have specialized machinery for creating both kinds of denotation, and we keep both of them around so that we can pick which one we want later.

Reply:
> I think this stuff probably needs a note (we'll hopefully delete it later but if we don't we'll wish we had the note).

It very much does, I did mention that I hadn't documented anything yet.

Reply:
Yeah sorry, I reviewed before I read all the final comment or something.